Engine API: add getTransitionConfiguration

mkalinin opened the pull request:
### Proposal
Proposes to add `engine_getTransitionConfigurationV1` method

### Motivation
* Allows for EL heartbeat and for verifying that EL client distribution is configured correctly
* Gives ability for CL to ping EL constantly, match received configuration with its own and surface a problem to a user if misconfiguration occurs

This proposal introduces a method that is useful only before the merge transition with intention to deprecate it in a subsequent upgrade. Low implementation complexity should mitigate this downside.

willianr2304 commented:
Como faço para receber

djrtwo commented on src/engine/specification.md:
```diff
@@ -310,3 +322,19 @@ The payload build process is specified as follows:
+#### Request
+
+* method: `engine_getTransitionConfigurationV1`
+* params: *empty list*
```
The alternative is that it's more of a ping/pong where each send their configuration. Thus both can see a mismatch and both can surface an error.

Do you think this would help reduce chance of user failures by both logs having the error?

ryanschneider replied:
+1 to this idea, basically `params` is the CL's configuration and `response` is the EL's, and both sides SHOULD log a warning if the other configuration doesn't match.

mkalinin replied:
I like this idea! 👍 I think we may re-use the same data structure. When `TERMINAL_BLOCK_HASH` is set CL client will send `terminalBlockNumber: 0` as a block number is not a part of CL's configuration.

mkalinin replied:
Although, what if it's called not by CL but by a script that checks every EL client in a cluster and then matches values with what are expected to be? Do we want to design for this use case? If yes, then we should make the parameter optional.

djrtwo replied:
A cluster/advanced test could also add the config as a param. I think it's not worth the complxity of making this optional

djrtwo reviewed (commented):
Looks good! One minor question/suggestion

mkalinin commented:
I've updated the method to `engine_exchangeTransitionConfigurationV1` with `transitionConfiguration` in params. Also, added "**SHOULD** surface an error..." for both CL and EL clients.

cc @ryanschneider @djrtwo

djrtwo commented on src/engine/specification.md:
```diff
@@ -307,6 +319,27 @@ The payload build process is specified as follows:
+
+* method: `engine_exchangeTransitionConfigurationV1`
+* params:
+  1. `transitionConfiguration`: `Object` - instance of [`TransitionConfigurationV1`](#TransitionConfigurationV1); `terminalBlockNumber` **MUST** be set to `0`
```
Why must `terminalBlockNumber` be set to 0?

couldn't the config be overridden to use TBH and TBN?

djrtwo replied:
TBN is always 0 on CL. Just TBH is used

yezawlin commented:
I'm sorry I made a mistake I do not know how to do it again  please help all


https://bit.ly/34l8flv



On Mon, 7 Feb 2022, 20:38 Danny Ryan, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/engine/specification.md
> <https://github.com/ethereum/execution-apis/pull/172#discussion_r800690084>
> :
>
> > @@ -307,6 +319,27 @@ The payload build process is specified as follows:
>
>  1. Given the `payloadId` client software **MUST** return the most recent version of the payload that is available in the corresponding build process at the time of receiving the call.
>
> -1. The call **MUST** return `-32001: Unknown payload` error if the build process identified by the `payloadId` does not exist.
> +2. The call **MUST** return `-32001: Unknown payload` error if the build process identified by the `payloadId` does not exist.
> +
> +3. Client software **MAY** stop the corresponding build process after serving this call.
> +
> +### engine_exchangeTransitionConfigurationV1
> +
> +#### Request
> +
> +* method: `engine_exchangeTransitionConfigurationV1`
> +* params:
> +  1. `transitionConfiguration`: `Object` - instance of [`TransitionConfigurationV1`](#TransitionConfigurationV1); `terminalBlockNumber` **MUST** be set to `0`
>
> TBN is always 0 on CL. Just TBH is used
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/execution-apis/pull/172#discussion_r800690084>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMRUL43SHDY2M4RGKXYUGKDUZ7G5ZANCNFSM5NLLAYAQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

ryanschneider reviewed (approved):
lgtm!
